File: Ising/ising2D.py
from random import randint, randrange, random
from math import exp
import matplotlib.pyplot as plt

N = 350
J = -1
H = 0
Kb = 1
trigger = 300000
sampleDelay = 500
res_arr = []
n_samples = 300
S = [[randrange(-1, 2, 2) for i in range(N)] for j in range(N)]

def runCycles(T, H = 0):
    m_arr_loc = []
    e_arr_loc = []
    for i in range(trigger + n_samples * sampleDelay):
        m_loc = 0
        e_loc = 0
        x = randint(0, N - 1)
        y = randint(0, N - 1)
        deltaE = -2 * J * S[x][y] * (S[(x + 1) % N][(y + 1) % N] +
                                  S[(x + 1) % N][(y - 1) % N] +
                                  S[(x - 1) % N][(y + 1) % N] +
                                  S[(x - 1) % N][(y - 1) % N]) - 2 * H * S[x][y]

        if deltaE < 0 or random() < exp(-deltaE/(Kb * T)):
            S[x][y] *= -1

        if i % sampleDelay == 0 and i > trigger:
            for j in S:
                for el in j:
                    m_loc += el
            m_arr_loc.append(m_loc / N**2)

            for a in range(N):
                for b in range(N):
                    e_loc += J * S[a][b] * (S[(a + 1) % N][b] + S[a][(b + 1) % N]) - 2 * H * S[a][b]
            e_arr_loc.append(e_loc)

    susc = 0
    sq_av = 0
    av_sq = 0
    magn = 0

    for m in m_arr_loc:
        sq_av += m / len(m_arr_loc)
        av_sq += m**2 / len(m_arr_loc)

    magn = abs(sq_av)

    sq_av = sq_av ** 2
    susc = (av_sq - sq_av) / T

    cal = 0
    sq_av = 0
    av_sq = 0
    for e in e_arr_loc:
        sq_av += e / len(m_arr_loc)
        av_sq += e**2 / len(m_arr_loc)

    sq_av = sq_av ** 2
    cal = (av_sq - sq_av) / T

    return [T, susc, cal, magn]

for t in range(400, 1, -4):
    res_arr.append(runCycles(t / 100))

susc = []
temp = []
heat = []
arr_magn = []
for i in range(len(res_arr)):
    temp.append(res_arr[i][0])
    susc.append(res_arr[i][1])
    heat.append(res_arr[i][2])
    arr_magn.append(res_arr[i][3])

plt.xlabel("Temperature")
plt.ylabel("Susceptibility")
plt.plot(temp, susc)
#plt.show()

plt.ylabel("Specific heat")
plt.xlabel("Temperature")
plt.plot(temp, heat)
#plt.show()

plt.ylabel("Magnetization")
plt.xlabel("Temperature")
plt.plot(temp, arr_magn)
#plt.show()

File: Ising/Ising2D.jl
N = 35
J = -1
H = 0
Kb = 1
trigger = 30000
sample_delay = 500
n_samples = 300

S = [[rand(-1:2:1) for i = 1:N] for j = 1:N]

function getΔE(x, y)
	lowerX = x - 1
	upperX = x + 1
	lowerY = y - 1
	upperY = y + 1
	if x - 1 == 0 lowerX = N
	elseif x + 1 > N upperX = 1 end
	if y - 1 == 0 lowerY = N
	elseif y + 1 > N upperY = 1 end
    ΔE = -2 * J * S[x][y] * (S[upperX][y] + S[x][upperY] +
                      S[lowerX][y] + S[x][lowerY])
    return ΔE
end

function run_cycles(T)
    m_arr_loc = []
    for i = 1 : (trigger + n_samples * sample_delay)
        m_loc = 0
        x = rand(1 : N)
        y = rand(1 : N)
        ΔE = getΔE(x, y)

        if (ΔE < 0) | (rand() < exp(-ΔE / (Kb * T)))
            S[x][y] *= -1
        end

        if (i % sample_delay == 0) & (i > trigger)
            for j = 1:N, k = 1:N
                m_loc += S[j][k]
            end
            append!(m_arr_loc, m_loc / N ^ 2)
        end
    end
	magn = 0
    for i = 1:n_samples
		magn += m_arr_loc[i]
	end
	magn /= n_samples
	magn = abs(magn)
	return magn
end

res_arr = []
temp_arr = []

function go()
	for t = 0.01:0.01:4
		append!(res_arr, run_cycles(t))
		append!(temp_arr, t)
	end
end

@time(go())

using Plots
plot(temp_arr, res_arr)
